### DISCUSSION TITLE
Replacing a jobs args and schedule

### DISCUSSION THREAD

**Discussion by pconstantinou** [riverqueue/river#124] [Ideas]
I want to insert a job and then override it with newer args and an updated schedule.

In the specific use case, I'd like to trigger a long-running remote process with some data. As soon as the data comes to my service, I trigger the job to start in 45 seconds, but if more data comes in, I want to update the arguments and only have the most recent queued item handled.

Does River have that capability? It looks like the first one wins and duplicate inserts are discarded.

**Comment by bgentry** [comment]
@pconstantinou There isn’t any kind of generalized job update API as of now. This also seems like a bit of a specialized use case so it may not be something we get to soon.

Based on what you’re describing, you could build a reasonably fast query if you’re filtering on top level args keys using the GIST index. You should be careful to do this transactionally by taking a lock for update on the row you want to update and only continuing if its status is still `scheduled`. Should be reliable and fairly straightforward.

**Reply by pconstantinou** [reply]
Thanks for the fast reply and the advice on the database locking. Making OpenAI can take 30+ seconds so I've benefitted from queuing and only submitting the prompt when the input is idle. A queuing system was a nice solution but I can see why you wouldn't want to do this in a database. I've got it in RAM now but wondering how many places I can replace in memory channels with queues to improve robustness.
